File: src/main/java/org/antop/java/chapter3/Join.java
package org.antop.java.chapter3;

import java.util.Collection;
import java.util.List;

public class Join {
    private static final String DEFAULT_SEPARATOR = ", ";
    private static final String DEFAULT_PREFIX = "(";
    private static final String DEFAULT_POSTFIX = ")";

    public static <T> String joinToString(Collection<T> collection) {
        return joinToString(collection, DEFAULT_SEPARATOR);
    }

    public static <T> String joinToString(
            Collection<T> collection,
            String separator
    ) {
        return joinToString(collection, separator, DEFAULT_PREFIX);
    }

    public static <T> String joinToString(
            Collection<T> collection,
            String separator,
            String prefix
    ) {
        return joinToString(collection, separator, prefix, DEFAULT_POSTFIX);
    }

    public static <T> String joinToString(
            Collection<T> collection,
            String separator,
            String prefix,
            String postfix
    ) {
        var sb = new StringBuilder(prefix);
        var i = 0;
        for (T e : collection) {
            if (i++ > 0) sb.append(separator);
            sb.append(e);
        }
        sb.append(postfix);
        return sb.toString();
    }

    public static void main(String[] args) {
        var list = List.of(1, 2, 3);
        var join = joinToString(list);
        System.out.println(join);
    }
}

File: src/main/kotlin/org/antop/kotlin/chapter3/join.kt
package org.antop.kotlin.chapter3

var opCount = 0

@JvmOverloads
fun <T> Collection<T>.joinToString(
    separator: String = ", ",
    prefix: String = "(",
    postfix: String = ")"
): String {
    val sb = StringBuilder(prefix)
    for ((i, e) in this.withIndex()) {
        if (i > 0) sb.append(separator)
        sb.append(e)
    }
    sb.append(postfix)
    return sb.toString()
}

fun main() {
    val list = listOf(1, 2, 3)
    val join = list.joinToString(
        postfix = ") -> end"
    )
    println(join)

}
